### PR TITLE
Early hints extension

### PR DESCRIPTION
**Pull request by pgjones**
This gives an ASGI mechanism for the application to indicate to the
server what early hints to send.

See https://developers.cloudflare.com/cache/about/early-hints/, https://httpwg.org/specs/rfc8297.html, and https://www.rfc-editor.org/rfc/rfc8288.html.

I hope to implement this in Hypercorn and replace my server push usage.

### CODE REVIEW

File: docs/extensions.rst [hunk 1]
```diff
@@ -133,3 +133,37 @@ TLS
 ---
 
 See :doc:`specs/tls`.
+
+Early hints
+-----------
+
+A informational response with the status code 103 is an early hint,
```
Comment (Kludex) [inline docs/extensions.rst:140]:
```suggestion
An informational response with the status code 103 is an early hint,
```

File: docs/extensions.rst [hunk 2]
```diff
@@ -133,3 +133,37 @@ TLS
...
+*Response Body* message.
+
+Keys:
+
+* ``type`` (*Unicode string*): ``"http.response.early_hint"``
+
+* ``links`` (*Iterable[byte string]*): An iterable of link header field
+  values, see ``RFC 8288``.
+
+The ASGI server should then attempt to send an infromational response
```
Comment (Kludex) [inline docs/extensions.rst:166]:
```suggestion
The ASGI server should then attempt to send an informational response
```

File: docs/extensions.rst (hunk 1 above)
Comment (pgjones) [reply docs/extensions.rst]: Thanks, fixed

File: docs/extensions.rst (hunk 2 above)
Comment (pgjones) [reply docs/extensions.rst]: Thanks fixed

**Comment by andrewgodwin** [comment]
Ah yes, with Server Push basically being officially dead it does seem like time. This looks pretty good - any concerns or other ways you might want to do it you avoided here?

**Comment by pgjones** [comment]
I had thought of a more general `http.informational_response` message, but I can't think of why the ASGI app should want to send any other types. I also considered being more prescriptive about the link structure, but I think that is something the ASGI app can structure and decide like other headers.

File: docs/extensions.rst [hunk 3]
```diff
@@ -133,3 +133,37 @@ TLS
 ---
 
 See :doc:`specs/tls`.
+
+Early hints
```
Comment (andrewgodwin) [inline docs/extensions.rst:137]: Let's capitalise Hints here, as that's how everyone refers to it

File: docs/extensions.rst [hunk 4]
```diff
@@ -133,3 +133,37 @@ TLS
...
+-----------
+
+An informational response with the status code 103 is an early hint,
```
Comment (andrewgodwin) [inline docs/extensions.rst:140]: Can we have Early Hint either capitalised or in quotes here to indicate it is the proper name of the HTTP feature?

**Review by andrewgodwin — COMMENTED** [review]
OK, sounds sensible - a couple pf tiny copy-edit changes and then I'm pretty happy with this.

File: docs/extensions.rst (hunk 3 above)
Comment (pgjones) [reply docs/extensions.rst]: Yep, done

File: docs/extensions.rst (hunk 4 above)
Comment (pgjones) [reply docs/extensions.rst]: Yep, done

**Comment by andrewgodwin** [comment]
Lovely, thanks. Let's hope this gains much more traction overall than Server Push!

**Pull request merged**